### PR TITLE
Add unit test for surfacing potential regression from #5728

### PR DESCRIPTION
**Pull request by ncooke3**
Adds a currently failing unit test to highlight one potential regression from #5728. 

I stumbled upon this when building up the test suite in #5919. 

### Motivation:

#5728 introduced using `@_implementationOnly` when importing `Foundation.Bundle` in the generated resource bundle accessor. For Swift packages that import Foundation without `@_implementationOnly`, this seems to cause the following warning:
```console
UtilsWithFoundationPkg/Sources/UtilsWithFoundationPkg/FooUtils.swift:1:8: error: 'Foundation' inconsistently imported as implementation-only
    import Foundation
           ^
    @_implementationOnly 

UtilsWithFoundationPkg.build/DerivedSources/resource_bundle_accessor.swift:1:35: note: imported as implementation-only here
    @_implementationOnly import class Foundation.Bundle
```

This is reproducible when using the latest 5.8 toolchain and running the unit test added in this PR.

It seems that, by using `@_implementationOnly` in the resource bundle accessor to import Foundation,  it causes a warning if another swift file in the module imports Foundation without it. Ideally, the `@_implementationOnly` in the resource bundle accessor should be ignored when doing this check.  More info w.r.t. the warning in this bug: https://github.com/apple/swift/issues/56093.

### Modifications:

Add a unit test that is currently failing but should pass. The proper implementation fix isn't obvious to me but I wanted to open this both for reproduction and tracking purposes.

### Result:

Assuming CI is using the 5.8 toolchain, it should highlight this regression.

cc: @xwu @neonichu

### CODE REVIEW

**Comment by xwu** [comment]
😔

I'm not sure this warning should be triggered at all when the imports do not match exactly... Or, we should have some way to silence it in code, defaulting to parens.

File: Tests/FunctionalTests/ResourcesTests.swift
Comment (ncooke3) [inline Tests/FunctionalTests/ResourcesTests.swift:121]: Starting a thread to discuss...

> I'm not sure this warning should be triggered at all when the imports do not match exactly... Or, we should have some way to silence it in code, defaulting to parens.

I did some researching and there is some good discussion in https://forums.swift.org/t/update-on-implementation-only-imports/26996/12 as to why the warning is triggered in the first place. What makes this tricky is that I do not think there is a way to pass a flag to silence only this warning.

**Comment by tomerd** [comment]
cc @neonichu

**Comment by neonichu** [comment]
Thanks @ncooke3, this is a good point. 

We could do an import scan of modules which are using resources and adjust the generated code based on whether they import `Foundation` themselves or not, though the existing facility we have for this in SwiftPM actually does not tell us whether the imports are implementation-only themselves (which the previous version of the generated didn't address, either).

Given where we are in the 5.8 schedule, we may want to disable this change for the 5.8 tools-version and move it to `vNext` for now.

File: Tests/FunctionalTests/ResourcesTests.swift
Comment (xwu) [reply Tests/FunctionalTests/ResourcesTests.swift:121]: Understand the rationale, but this issue points to a use case where the current design doesn’t serve our needs (and, incidentally, if it were ever stabilized as an “access level” on `import` like once pitched, wouldn’t behave like access modifiers do elsewhere).

I posted some thoughts over in the linked PR about why even import scanning would be brittle, and I think possibly the most resilient solution here is to come up with a new “flavor” of `@_implementationOnly` (rather than a flag) that explicitly acknowledges that it can be overridden or “upgraded.”

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by neonichu** [comment]
@swift-ci smoke test macOS

**Comment by MaxDesiatov** [comment]
Are there any other updates needed for this PR?

**Comment by ncooke3** [comment]
@MaxDesiatov, nothing from my end.

**Comment by MaxDesiatov** [comment]
@swift-ci smoke test

**Comment by ncooke3** [comment]
@MaxDesiatov, good to merge?

**Comment by MaxDesiatov** [comment]
This can't be merged until it has approvals from reviewers.

**Comment by neonichu** [comment]
LGTM, but I wanted to run the test locally before approving since the `FunctionalTests` are still disabled.

**Comment by neonichu** [comment]
Sorry for dropping the ball on this, functional tests are back enabled now, so I am just going to run the CI again here.

**Review by neonichu — APPROVED** [review]

**Comment by ncooke3** [comment]
> Sorry for dropping the ball on this, functional tests are back enabled now, so I am just going to run the CI again here.

No worries– the main point of this PR was to highlight/demonstrate the issue! 🙂 And that's great about the functional tests, I'll get my mixed language branch synced up so the functional tests I added can run on CI.

**Comment by neonichu** [comment]
@swift-ci smoke test linux

**Pull request merged**
